### ISSUE TITLE
When is embed.fth invoked to set up the VM?

### ISSUE THREAD

**Issue by darozak** [Darwin-River/Ex-Machinis#52]
The game's Embed interpreter uses a file named [embed.fth](https://github.com/Darwin-River/Ex-Machinis/blob/develop/game-engine/src/embed/embed.fth) to set up the virtual machine and build FORTH's core dictionary.  This is where we will define many of the basic FORTH commands that will be used by players when programming their drones.

However, if I make a change to embed.fth, I'm not sure when that change will be folded into the game. This could occur globally, whenever we recompile the Game Engine or restart the game.  It might also occur every time a player sends a <reset> command to their drone.

I'll need to run a series of tests to figure out when this occurs.

**Comment by crodnun** [comment]
Hi Dave, just in case this could help you, here you have the explanation of the VM mechanics:

- We have a VM template for everybody that is the result of compiling the embed.fth you mention.
- Every user starts with this VM.
- The engine knows where to pick it using configuration, at game-engine.conf you can find this parameter:
#---------------------------
# FORTH configuration
#---------------------------
FORTH_IMAGE_PATH="/home/forth/game-engine/bin/eforth.blk"

This is the default VM that every user will start with.

We just load the binary version obtained after compiling the embed.fth.

**Comment by darozak** [comment]
Ok. That makes sense.  When and how do we compile embed.fth into eforth.blk? I would like to be able to generate an eforth.blk file whenever I need to so I can implement changes made to embed.fth.

**Comment by crodnun** [comment]
Hi @darozak , I have reviewed the old embed repo, and seems that Howard describes a bit how you can generate new images on your own.

I let you here the link from the author (look at Program Operation section):

[embed](https://github.com/howerj/embed)

I know how to do this stuff at C/compilers side but a bit confused with the FORTH's terminology, metacompilers, etc,...

As far as I understand, we should be able to modify the .fth file (meta-compiler), then generate a new block file just using the commands he describes there (this is easy for me, no problem there!!).

The tricky part is the .fth metacompiler modification, lol. If you know how to do it, generating a new version of the default VM will be easy. I know how to do the rest of the steps.

**Comment by darozak** [comment]
Would you be able to generate an executable that could be run from the shell or even in windows to convert a .fth file to a .blk file?

**Comment by crodnun** [comment]
Hi Dave,
For the moment the easiest for me is to generate the linux/shell version using Howard's code.

I put this program at advolition server/forth user, here:

```
[forth@advolition] # pwd
/home/forth/game-engine/test/dave
```
Also put a .fth file with the definitions (embed.fth).

Using both you can generate/meta-compile new VM this way:

```
[forth@advolition] # ./embed -o newvm.blk embed.fth 
FORTH META COMPILATION START 
COMPILATION COMPLETE
HOST:    3F48
TARGET:  1528
HEADER: 
  5000:     14     0  7FFF  2400   347  8000     0    14  _______$G_______
  5010:      0  7FFF  2400  4689  4854   A0D   A1A  1528  _____$_FTH____(_
  5020:   4DEA     1  1984     1   949  628D  601C  628D  _M______I__b_`_b
  5030:   631C  10E0  151E     0  7003  6461  4017  4100  _c_______pad_@_A
SAVING...DONE
STACK> <sp
```

Following the instructions described (here)[https://github.com/howerj/embed].

If you are interested in a windows version (embed.exe), I think that should not be a big deal, never done before but seems that it works there too (codes seem prepared to have an embed.exe).

**Comment by darozak** [comment]
Thanks Carlos,  This will be perfect.  I assume that, if I replace the blk file located at /home/forth/game-engine/bin/eforth.blk, the new base will appear in drones they are <reset>?

**Comment by crodnun** [comment]
This new VM will ve used for all the drones that do not have it in DB. This is the base one, if a user already has one in use, we do not update it.
What I normally do is to set VM = NULL for everybody (at agents table) and then any new user will pick the new base when using the game for the first time.

**Comment by crodnun** [comment]
What do you mean with meta command? The VM extensions like "query"? If so, the answer is no. So far the VM is created only once per user, when he/she issues the first command,
then it is stored/updated at DB

**Comment by darozak** [comment]
We addressed the <reset> meta command in the following issue:
https://github.com/Darwin-River/Ex-Machinis/issues/41

Based on what you've written there and above, it appears that a reset would force the VM to install the modified blk file.

**Comment by crodnun** [comment]
Ok understood, yes every time we reset the vm, the base one is picked to start from scratch.

**Comment by darozak** [comment]
Carlos,  I tried the following without success:

1. [Updated ](https://github.com/Darwin-River/Ex-Machinis/commit/2ed2473df8be2a1c135ff1e8dd3037838b9691da)embed.fth to include a new command that will add two to a number:
```
: 2+ 2 + ;
```
2. Uploaded the file to /home/forth/game-engine/test/dave and ran embed.
![Capture of embed command](https://user-images.githubusercontent.com/33500329/72668689-76240980-39f7-11ea-9e5a-774bc6c7f3d7.PNG)

3, Transferred the resulting eforth.blk into the root's bin directory:
![Capture of eforth blk upload](https://user-images.githubusercontent.com/33500329/72668738-df0b8180-39f7-11ea-97be-15147e17ec15.PNG)

4. Reset my drone using \<reset\> and received a message confirming that the VM was successfully reset.

5. Tried executing the following command:
```
<run>
1 2+ .
</run>
```

6. Received an error indicating that the word (2+) is not defined:
```
---- Output ----

1 2+ . -13?
```

Can you see any reason why the new command wasn't available in the base directory after I reset the drone?

**Comment by crodnun** [comment]
Hi Dave, I did the same test at my side and seems that it does not work as expected. Do you want I talk with Howard to confirm if we are doing this correctly? Seems that the new word is not present at the VM. The rest of words seem to be there but when you add a new one, no luck.

**Comment by darozak** [comment]
Hi Carlos.  Thanks for the guidance on avoiding the root!  I'll avoid using it and stick with forth.

That would be great if you could follow up with Howard on the embed application.  We'll need to modify the blk file to bring it in line with the game.  It will be useful to have a number of game-related definitions pre-loaded in the drones.

Thanks for your help and programming wizardry!

**Comment by crodnun** [comment]
Hi Dave, created an issue at Richard's project (yes, his name was Richard... not Howard, ... lol). Let's see if he can help us with this.

[issue](https://github.com/howerj/embed/issues/7)

**Comment by darozak** [comment]
Thanks Carlos!  It looks like Richard has been quick to respond as usual :)

**Comment by crodnun** [comment]
Hi Dave, Richard's feedback [here](https://github.com/howerj/embed/issues/7#issuecomment-577900944)

He has detected what is the problem and give us different alternatives to solve it. Feel free to review them and let me know what approach you would like to take.

**Comment by darozak** [comment]
This is awesome!  Richard provided a very nice description of what's happening.  I agree that we can loose a lot of unwanted words from the base image to make room for our own.   The second solution he offers would also work, if necessary.  

As always, I'll let you know if I run into any problems while playing around with this.  I'm looking forward to playing around with the code!  (I'll be careful not to break it :-)

Thanks for everything!!
